language/move-vm/runtime/src/native_functions.rs

```diff
@@ -8,23 +8,27 @@ use crate::{
 use move_binary_format::errors::{ExecutionState, PartialVMError, PartialVMResult};
 use move_core_types::{
     account_address::AccountAddress,
-    gas_schedule::CostTable,
     identifier::Identifier,
     language_storage::TypeTag,
     value::MoveTypeLayout,
     vm_status::{StatusCode, StatusType},
 };
 use move_vm_types::{
-    data_store::DataStore, gas_schedule::GasStatus, loaded_data::runtime_types::Type,
-    natives::function::NativeResult, values::Value,
+    data_store::DataStore, loaded_data::runtime_types::Type, natives::function::NativeResult,
+    values::Value,
 };
 use std::{
     collections::{HashMap, VecDeque},
     fmt::Write,
+    sync::Arc,
 };
 
-pub type NativeFunction =
-    fn(&mut NativeContext, Vec<Type>, VecDeque<Value>) -> PartialVMResult<NativeResult>;
+pub type UnboxedNativeFunction = dyn Fn(&mut NativeContext, Vec<Type>, VecDeque<Value>) -> PartialVMResult<NativeResult>
+    + Send
+    + Sync
+    + 'static;
+
+pub type NativeFunction = Arc<UnboxedNativeFunction>;
 
 pub type NativeFunctionTable = Vec<(AccountAddress, Identifier, Identifier, NativeFunction)>;
```

Comment:
I do not understand this change. I was expecting to see something much simpler. Just 
```
pub type NativeFunctionTable = Vec<(AccountAddress, Identifier, Identifier, NativeFunction, GasCost)>;
```

And then we could add the `GasCost` to the `NativeContext`

Reply:
Good question. My original plan was to add a single `GasCost`, but I ended up implementing a more generalized version based on the following reasoning: 
1. Only a package knows how its gas formulae can be customized and what gas parameters are needed, so ideally, the package themselves should have the freedom (and responsibility) to provide such customization to its own clients, in ways that make sense to the specific algorithms its native functions use.
2. From the Move VM's perspective, we really don't care how a native function wants to meter its own gas usage. The current protocol just requires it to return a number. Mandating the usage of a single `GasCost` is both restrictive and adds unnecessary coupling. I think in the general case, it's quite reasonable for a native function to have more than 1 gas parameter and for different native functions to have different number of gas parameters, potentially with different types.

Reply:
I'd prefer to keep this a bit simpler unless you have a hard use case. Reading this I feel like it really complicates the code. I get that it is "more general", but I feel like our goal should be to serve the common use cases we have. Really gnarly use cases can always dive into `NativeExtensions` and stuff gas data into that.

I don't understand this unnecessary coupling comment quite fully. But if you want something a little bit more flexible...
Given `GasCost` is `pub struct GasCost { pub instruction_gas: InternalGasUnits<GasCarrier>, pub memory_gas: InternalGasUnits<GasCarrier>, }`

We could do something like `pub struct NativeGasCost { pub instruction_gas: Option<InternalGasUnits<GasCarrier>>, pub memory_gas: Option<InternalGasUnits<GasCarrier>>, }`

Reply:
Pulling in @davidiw and @zekun000 / @msmouse; also interested in @sblackshear opinion.

I agree with @tnowacki that this looks complex at the first sight. However, I think it is easier to use then apparent from this PR. That is because the design pattern Victor choose is clear and easy to follow.

Basically, if you define a native, you also have to define a data structure for representing the parameters for the costing. Like here:

```
pub struct MyNativeGasParameters {
    pub base_cost: u64,
    pub unit_cost: u64,
}

fn my_native(
    gas_params: &MyNativeGasParameters,
    _context: &mut NativeContext,
    _ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    ...
    let cost = gas_params.base_cost + gas_params.unit_cost * <size of data>;
    NativeResult::map_partial_vm_result_one(cost, ...)
}
```

Then when you integrate this as a new native into a Move package, you do something like

```
struct PackageGasParameters {
   // existing parameters
   ...
   my_native_params: MyNativeGasParameters
```

I feel confident that this does not make the introduction of new functions significantly more difficult than it already is. 

The problem this solves, really, is that the computation of the cost of some operations cannot be done upfront just on basis of the input parameters. At least, not efficiently -- computing the cost can be as expensive as doing the operation. Consider for example a table read. We may want to cost a table read differently based on whether it is the first or subsequent access in a given txn. So instead of hacking around this for tables and doing it somehow over the native context extension, Victor's design provides a general clean model for costs of natives which can be applied to any current and potential future use cases.

Reply:
I hear your concerns. Let me explain my thoughts from the perspective of each individual component:
1. For the Move VM
  a. The point of the decoupling is to give the control of gas to the individual crates, and it's up to them to choose how they want to customize their gas formula. The VM supports all options: complex, simple, or even non-customizable at all. This way the VM's own logic remains simple. I understand your argument that passing in a single GasCost would cover a lot of use cases, but it still feels a bit like a leaky abstraction to me. (My current move-stdlib implementation opted for the complex path and probably left you the bad impression that this whole thing is inherently complex. I'll address this later)
  b. The change from `fn(...)` to `Arc<dyn Fn(...)>` actually results in very minimal changes to the VM code itself -- only a handful lines were are to this. In terms of performance, this does add a virtual function call, but I highly doubt if this is noticeable at all. The `Arc` is only used to make the existing loader implementation happy and satisfy its questionable assumption that the function needs to be accessed by multiple threads.
  c. I think you made a good point about `NativeExtensions`. The table extension for example, gets its gas parameters this way. My concern about this however, is that this prevents code inlining and other compiler optimizations compared to the closure approach.
2. For move-stdlib (or other packages with native functions)
  a. As I hinted above, it is my current implementation for this package that opted for the more complex path, defining gas parameters for every function and every upper layer (module/crate). This does grant us maximum composability but I also feel like it's a bit too cumbersome. I would be open to switch to a simpler model. (e.g. `fn make_natives(address: AccountAddress, gas_costs: &[GasCost])` + getting rid the individual gas parameter structs and makers)
  b. See `examples/diem-framework/natives` for a simpler (actually now non-configurable) example. There is very little added code to make it work with the new system. Again, I would argue that the complexity is really an issue on the package side -- in general, the more composable/customizable you want your package to be, the more complex your boilerplate code will be.

Reply:
> I would be open to switch to a simpler model. (e.g. fn make_natives(address: AccountAddress, gas_costs: &[GasCost]) + getting rid the individual gas parameter structs and makers)

I feel not good with this. This is optimizing friction at the wrong level of abstraction. We aren't designing APIs for end users.

We should make this canonical and maximally expressive. There is some boilerplate, but its easy to create as long as you have some existing code to start from.

Reply:
FYI, I did try to write a macro that would save us from having to write the boilerplate code and it worked:
```
define_native!{
    name = foo;

    gas_params = {
        base_cost: u64,
    }

    (gas_params, context, ty_args, args) => {
        ...
    }
}
```
The problem however, is that we lose move-analyzer's auto-formatting and type annotation, which made me give this up.

Reply:
@wrwg 
> I feel confident that this does not make the introduction of new functions significantly more difficult than it already is.

I tend to agree. My only observation was that most of these structs are more or less just a `GasCost`. And in the cases they aren't, we could maybe get away with a similar version of `GasCost` but where each field is an `Option`.

I guess what I'm trying to say is: it would be nice if we don't have to define these structs (and from looking at the examples it feels a bit like overkill), but it is not the end of the world if we have to.

Reply:
The `PackageGasParameters` is easier to read than cost_table, but how we should keep the `PackageGasParameters`? On Starcoin, we keep the `cost_table` to Move module.

https://github.com/starcoinorg/starcoin-framework/blob/d3bb99ff026e408341377ecd73ce37108c278066/sources/VMConfig.move#L224-L233

This approach makes it easier to add a new native method or update the gas cost of a native method.

If the GasParameters is a struct, it is harder to extend than vector.

Reply:
@jolestar this is actually a very good question. 
1. The reason why I opted for a PackageGasParameters struct is because I wanted to have different gas parameters for different native functions.
2. I should make it clear that this struct is only intended to be used as an interface to configure the move-stdlib natives. You don't necessarily have to use it in other parts of your program or store it on-chain. You can and are encouraged to use your own format if you find that more convenient. It might be helpful to write a conversion function from your format to PackageGasParameters. We are also planning to do something similar in the Aptos framework: we'll define our own on-chain gas schedule format, which will likely be vector-based so it can be extended in the future.